### PR TITLE
Add callback argument

### PR DESCRIPTION
**Pull request by dotcypress**
Add callback argument to build some cool stuff around `linkstate` 👌

We use our bike-shed for form validation, but our implementation of `linkState` is so naive and we really want to use this great module as dependency. 👍

### CODE REVIEW

**Comment by developit** [comment]
Hey @dotcypress - good call on this.  Just so I have a better grasp in my head, what's your best _"here's why this is awesome"_ use-case for the callback?

> context: I ask because I have long debated adding a "mutator" argument to linkState(), which would be given the opportunity to modify the state update object prior to it being passed to `setState()`.  However, I never ended up having a need for it - I generally store raw data in state and derive from it in `render()`, so directly linking DOM events into state has always been sufficient.  If others see that the same way, then a callback argument like you've petitioned for here is the natural next step 👍

File: src/index.js [hunk 1]
```diff
@@ -4,9 +4,10 @@ import delve from 'dlv';
  *	@param {Component} component	The component whose state should be updated
  *	@param {string} key				A dot-notated key path to update in the component's state
  *	@param {string} eventPath		A dot-notated key path to the value that should be retrieved from the Event or component
+ *	@param {function} cb			A function to be called once component state is updated
  *	@returns {function} linkedStateHandler
  */
-export default function linkState(component, key, eventPath) {
+export default function linkState(component, key, eventPath, cb) {
```
Comment (developit) [inline src/index.js]: Only suggestion I'd make would be to use the full word `callback` here.  It gets minified out anyway, and it's just that little bit more understandable for newcomers! 😊

**Comment by dotcypress** [comment]
Hey @developit 
I will show my thoughts with example code. `linkedForm` is tiny form validator. Looks weird, but i really like experiments. 😄

```js
import linkState from 'linkstate'
const noop = () => {}

export default function linkedForm (component, rules, opts = {}) {
  let pristine = true

  function createPatch (errors = {}) {
    return { [opts.errorsPath || 'errors']: errors }
  }

  function validate () {
    pristine = false
    const errors = Object.keys(rules).reduce((acc, statePath) => {
      const validator = rules[statePath] || noop
      const error = validator(component.state)
      if (error) {
        acc[statePath] = error
      }
      return acc
    }, {})
    component.setState(createPatch(errors))
    return Object.keys(errors).length === 0
  }

  Object.assign(component.state, createPatch())

  return {
    validate,
    linkState: (statePath, valuePath) =>
      linkState(component, statePath, valuePath, () => !pristine && validate()),
    reset: () => {
      pristine = true
      component.setState(createPatch())
    }
  }
}
```

Example component:

```js
class SignupPage extends Component {
  constructor (props) {
    super(props)
    this.state = {
      email: '',
      password: '',
      firstName: '',
      lastName: ''
    }
    this.form = linkedForm(this, {
      firstName: ({ firstName, mode }) => !firstName && 'Please enter a valid name.',
      lastName: ({ lastName, mode }) => !lastName && 'Please enter a valid name.',
      email: ({ email }) => !isEmail(email) && 'Please enter correct email',
      password: ({ password }) => password.length < 5 && 'Password must contain at least 5 characters.'
    })
  }

  handleSubmit () {
    const { email, password, firstName, lastName } = this.state
    if (!this.form.validate()) {
      return
    }
    signup(email, firstName, lastName, password).then(() => this.form.reset())
  }

  render () {
    const { errors, email, password, firstName, lastName } = this.state
    return (
      <form onSubmit={this.handleSubmit}>
        <input type='text' value={firstName} onChange={this.form.linkState('firstName')}/>
        {errors.firstName && <strong>{errors.firstName}</strong>}
        <input type='text' value={lastName} onChange={this.form.linkState('lastName')}/>
        {errors.lastName && <strong>{errors.lastName}</strong>}
        <input type='email' value={email} onChange={this.form.linkState('email')}/>
        {errors.email && <strong>{errors.email}</strong>}
        <input type='password' value={password} onChange={this.form.linkState('password')}/>
        {errors.password && <strong>{errors.password}</strong>}
        <button type='submit'>Register</button>
      </form>
    )
  }
}
```

As you see, we need to revalidate form in some cases, but we use current component state as input. So, we need to wait for state settle. Of course, i can validate input before placing it to state, and validation result will be saved to state. But in this situation, validation changes renders before actual state changes, and this behavior is questionable.

**UPD:** And, well, if i will validate before calling `setState`, i need to extract data from event and map it to state(sounds like `linkstate` functionality, huh 😄 )

**Review by developit — APPROVED** [review]

**Comment by developit** [comment]
Nice - I'm seeing hints in there that you've done Angular haha.

Question though - in the above example, could validation have been done in `componentDidUpdate()`?  I guess it might be hard to avoid accidentally creating an infinite render loop... might have just answered my own question there.

**Comment by dotcypress** [comment]
Angular? It was long time ago 😆 

Yes, `componentDidUpdate` is dangerous place for updating state.

**Comment by dotcypress** [comment]
Kindly reminder on this.

**Comment by developit** [comment]
Yup, sorry for dropping this on the floor. Still need to contact a few people about this to make sure we're good to change the API.

**Comment by paranoidjk** [comment]
+1 for this

**Pull request closed without merging**